crates/router/src/core/payouts/validator.rs

```diff
@@ -317,3 +331,71 @@ pub async fn create_payout_link_db_entry(
             message: "payout link already exists".to_string(),
         })
 }
+
+pub fn validate_payout_link_render_request(
+    request_headers: &header::HeaderMap,
+    payout_link: &PayoutLink,
+) -> RouterResult<()> {
+    let link_id = payout_link.link_id.to_owned();
+    let link_data = payout_link.link_data.to_owned();
+
+    // Fetch destination is "iframe"
+    match request_headers.get("sec-fetch-dest").and_then(|v| v.to_str().ok()) {
+        Some("iframe") => Ok(()),
+        Some(requestor) => Err(report!(errors::ApiErrorResponse::AccessForbidden {
+            resource: "payout_link".to_string(),
+        }))
+        .attach_printable_lazy(|| {
+            format!(
+                "Access to payout_link [{}] is forbidden when requested through {}",
+                link_id, requestor
+            )
+        }),
+        None => Err(report!(errors::ApiErrorResponse::AccessForbidden {
+            resource: "payout_link".to_string(),
+        }))
+        .attach_printable_lazy(|| {
+            format!(
+                "Access to payout_link [{}] is forbidden when sec-fetch-dest is not present in request headers",
+                link_id
+            )
+        }),
+    }?;
+
+    // Validate origin / referer
+    let domain_in_req = request_headers.get("origin")
+    .or_else(|| request_headers.get("referer"))
+    .and_then(|v| v.to_str().ok())
+    .ok_or_else(|| report!(errors::ApiErrorResponse::AccessForbidden {
+        resource: "payout_link".to_string(),
+    }))
+    .attach_printable_lazy(|| {
+        format!(
+            "Access to payout_link [{}] is forbidden when both origin and referer headers are missing from the request headers",
+            link_id
+        )
+    })?;
+
+    if is_domain_allowed(domain_in_req, link_data.allowed_domains) {
+        Ok(())
+    } else {
+        Err(report!(errors::ApiErrorResponse::AccessForbidden {
+            resource: "payout_link".to_string(),
+        }))
+        .attach_printable_lazy(|| {
+            format!(
+                "Access to payout_link [{}] is forbidden from requestor - {}",
+                link_id, domain_in_req
+            )
+        })
+    }
+}
+
+fn is_domain_allowed(domain: &str, allowed_domains: HashSet<String>) -> bool {
+    allowed_domains.iter().any(|allowed_domain| {
+        Regex::new(allowed_domain)
```

Comment:
You cannot do this, allowed domains are not valid regex patterns in the first place. If you consider `*.example.com`, this is an invalid regex pattern.
Secondly, I think accepting regexes as user input is a bad idea.